core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java

```diff
@@ -203,24 +208,47 @@ private void assertValidBuddy(UserDto requestingUser, BuddyDto buddy)
 	}
 
 	@Transactional
-	public BuddyDto addBuddyToAcceptingUser(UserDto acceptingUser, UUID buddyUserId, String buddyNickName,
-			UUID buddyUserAnonymizedId, boolean isRequestingSending, boolean isRequestingReceiving)
+	public BuddyDto addBuddyToAcceptingUser(UserDto acceptingUser, BuddyConnectRequestMessage connectRequestMessageEntity)
 	{
 		if (acceptingUser == null)
 		{
 			throw BuddyServiceException.acceptingUserIsNull();
 		}
+		if (!connectRequestMessageEntity.getSenderUser().isPresent())
+		{
+			throw UserServiceException.notFoundById(connectRequestMessageEntity.getSenderUserId());
+		}
 
 		acceptingUser.assertMobileNumberConfirmed();
-		Buddy buddy = Buddy.createInstance(buddyUserId, buddyNickName,
-				isRequestingSending ? Status.ACCEPTED : Status.NOT_REQUESTED,
-				isRequestingReceiving ? Status.ACCEPTED : Status.NOT_REQUESTED);
-		buddy.setUserAnonymizedId(buddyUserAnonymizedId);
+		Buddy buddy = Buddy.createInstance(connectRequestMessageEntity.getSenderUser().get().getId(),
+				connectRequestMessageEntity.getSenderNickname(),
+				connectRequestMessageEntity.requestingSending() ? Status.ACCEPTED : Status.NOT_REQUESTED,
+				connectRequestMessageEntity.requestingReceiving() ? Status.ACCEPTED : Status.NOT_REQUESTED);
+		buddy.setUserAnonymizedId(connectRequestMessageEntity.getRelatedUserAnonymizedId().get());
+		createBuddyDevices(connectRequestMessageEntity).forEach(buddy::addDevice);
 		BuddyDto buddyDto = BuddyDto.createInstance(buddyRepository.save(buddy));
 		userService.addBuddy(acceptingUser, buddyDto);
 		return buddyDto;
 	}
 
+	private Set<BuddyDevice> createBuddyDevices(BuddyConnectMessage connectMessageEntity)
+	{
+		List<String> deviceNames = connectMessageEntity.getDeviceNames();
+		List<UUID> deviceAnonymizedIds = connectMessageEntity.getDeviceAnonymizedIds();
+		List<Boolean> deviceVpnConnectionStatuses = connectMessageEntity.getDeviceVpnConnectionStatuses();
+		int numDevices = deviceNames.size();
+		assert deviceAnonymizedIds.size() == numDevices;
+		assert deviceVpnConnectionStatuses.size() == numDevices;
+
+		Set<BuddyDevice> devices = new HashSet<>();
+		for (int i = 0; (i < numDevices); i++)
```

Comment:
Shouldn't we use streams instead of for loops?

Reply:
I woud like to, but I don't know how we could do that. In Groovy, you could do ``eachWithIndex``, but Java doesn't have that.

Reply:
Could maybe be done using `IntStream.range(0, numDevices).foreach(...)`? See [here](https://stackoverflow.com/questions/18552005/is-there-a-concise-way-to-iterate-over-a-stream-with-indices-in-java-8).

Reply:
I have seen that, but that doesn't make it more readable.